### ISSUE TITLE
Global suggestion: don't use lib/ as a top-level directory

### ISSUE THREAD

**Issue by aantron** [tmattio/dream-cli#3]
If newcomers use these repos as starting points, they might think `lib/` is a good idea, and then be bitten when ReScript assumes `lib/` is open for its own internal usage.

I personally don't know where this convention even comes from; at least in C-like environments `lib/` would only ever be an output directory, not something that appears in the repo, except maybe after build. That is also consistent with how ReScript uses it. Sources should not be under top-level `lib/`, a directory traditionally for output or intermediate files.

Another drawback of having multiple top-level "source" directories (as opposed to grouped under `src/`), as often follows from `lib/` (e.g. then also `bin/`, `lib_lwt/`, etc.) is that you have to list all of them if adding a recursive watcher. With `src/`, one just watches `src/` (and maybe `test/`).

**Comment by aantron** [comment]
I've only ever really seen top-level `lib/` sources in Mirage-style projects, but it's arguably wrong, and only "gliding by" because Mirage projects are not typically using ReScript, or autotools, or other tools assuming more "standard" conventions.

**Comment by tmattio** [comment]
Thanks for the suggestion!

While I appreciate that there are some drawbacks to using `{bin,lib}/` and I don't have a strong opinion on what directory structure to use, I am a strong believer in following the ecosystems conventions for this kind of decision.

Elixir is a good example of an ecosystem where the directory structure convention is almost universally adopted, and this is very powerful for beginners: they can read one tutorial about Elixir conventions, and be able to navigate any Elixir project!

In OCaml, there's, unfortunately, no obvious winning convention, and this causes projects to follow many different conventions, which I believe is detrimental for beginners.

This is something I'd like to influence. Spin in particular, if it becomes widely used, is a good candidate to push the ecosystem toward a given convention, and let everybody reap the benefits this will bring.

With that being said, what prompted me to chose `{bin,lib}/` as a convention for Spin and the other projects I work on is the trend there seems to be for this directory structure in the ecosystem. Indeed, even if there is no real winners and you can see all sort of project structure, `{bin/lib}/` is used by some popular repositories of the OCaml Platform, especially the younger ones:

- [Dune](https://github.com/ocaml/dune)
- [Dune Release](https://github.com/ocamllabs/dune-release)
- [ocamlformat](https://github.com/ocaml-ppx/ocamlformat)
- [mdx](https://github.com/realworldocaml/mdx)

To give a full picture, `opam` and `odoc` are examples of projects that use `src/` as a convention.

One benefit of using `{bin,lib}/` is also to separate the definition of the CLI from the rest of the code. When landing on a repository that has a CLI (especially one that uses `cmdliner`), you can be almost sure that `bin/` will only contain the documentation and term definitions, and the actual business logic will be located in `lib/`.

So, again, I don't have an opinion on which one is better, there seem to be pros and cons to both conventions. But I'd like to help the ecosystem to establish a convention and a good way to do this seems to be to use the best candidate (measured very scientifically in number of repos?).

I'd be more than open to change the convention I use in either of these scenarios:

- There is a clear pain point to using `lib/`
- The ecosystem seems more likely to adopt another convention
- We can collectively (the OCaml community) agree to use a different convention

**Comment by aantron** [comment]
To add to the above, all of my projects personally are "younger" and very consciously and deliberately use `src/`, as well as older projects I maintained, which used `src/` and kept using it very intentionally. So...:

- [Lwt](https://github.com/ocsigen/lwt)
- [Bisect_ppx](https://github.com/aantron/bisect_ppx)
- [Lambda Soup](https://github.com/aantron/lambdasoup) and [Markup.ml](https://github.com/aantron/markup.ml), which are Dream-adjacent.
- [Dream](https://github.com/aantron/dream) itself, most relevantly.
- [odoc](https://github.com/ocaml/odoc), which you already mentioned.

I didn't do a search over all projects, so I'm just listing what I immediately know and is somehow relevant, hence projects I am familiar with.

> * There is a clear pain point to using `lib/`

There is a very clear pain point to Dream-related projects using `lib/` &mdash; as I mentioned, this conflicts with ReScript (not sure about Melange at this point) expecting to be able to use `lib/` as an intermediate files directory. This is no fault of ReScript, but is in accordance with long-established convention that a subset of the OCaml community did not heed.

> One benefit of using `{bin,lib}/` is also to separate the definition of the CLI from the rest of the code.

This benefit is not specific to `bin`/`lib`, as you can have `src/lib/` and `src/bin/`, and basically all substantial projects do something like that (e.g. [odoc](https://github.com/ocaml/odoc/tree/master/src/odoc/bin), [Bisect_ppx](https://github.com/aantron/bisect_ppx/tree/master/src)) without having to commit to `bin/` (or `lib/`) being at the top level of the directory tree.

Other disadvantages of `lib/` and `bin/` at top level, besides the `lib/` conflict:

- A larger list of directories to watch.
- A more messy top-level structure, forcing readers of a complicated project with many optional sub-libraries to wade through all of them looking for the non-sources.
- A taller repository file list, which pushes down the README.
- If a project has multiple sublibraries, a commitment to top-level directories forces strange things like `lib_lwt`, which quickly multiply causing a mess. They should have been all under a subdirectory to begin with, like `src/` for the core, `src/lwt/` for the Lwt stuff.
- What if a project has multiple binaries, like Bisect_ppx? Do we now have `bin_ppx` and `bin_report`? It looks like that underscore should be a slash, again strongly suggesting that subdirectories are the way to go. I understand this issue is not specifically about subdirectories, but the argument is that (1) there shouldn't be a top-level `lib/`, (2) all the benefits of a top-level `lib/` are better achieved using subdirectories of `src/` (or other favorite name) as necessary. You can still have all the right separate directories under `src/`.

I would be wary of statements about "young" and "popular' projects as grounds for anything, as there are many "young" projects following all sorts of conventions, including questionable retrogressive formatting conventions based on the drawbacks of other languages (e.g. Haskell and Elm, not saying they has *only* drawbacks, OCaml has its own *other* drawbacks) and older code bases, "novel" directory structures that haven't actually been tested against environments (ReScript), non-accessible usage of colors to represent distinctions (GitHub in a recent discussion we had), etc. Other "young" projects are avoiding these mistakes (but may well have yet other mistakes that they are making &mdash; no claim of overall superiority, just going issue by issue).

So I don't think "younger" or "popular" matters so much &mdash; bad conventions get picked up and established in environments that don't challenge them, etc. Again, not suggesting that projects that get *this* issue "right" according to me, don't suffer from the same problem in other respects according to me or others.

> So, again, I don't have an opinion on which one is better, there seem to be pros and cons to both conventions.

I'm only aware of one pro so far, which is a flatter directory structure is more immediately parsable in terms of "what is everything that is in this repo." However, it can almost equally be well hidden behind only one level of indirection (`src/`), or retained flat, just without a directory named `lib/`, which is the specific point that will tend to bite some people who do Web development.

> * We can collectively (the OCaml community) agree to use a different convention

The ReScript community has already agreed to use a convention where `lib/` is not a source directory. As I said, I think that's fine, and the novelty of `lib/` as a top-level source directory was always something a bit likely to run into trouble eventually.

**Comment by aantron** [comment]
> I didn't do a search over all projects, so I'm just listing what I immediately know and is somehow relevant, hence projects I am familiar with.

I just recalled, Daniel Buenzli's libraries are quite popular, and they seem to follow the "standard" `src/` convention and do it quite beautifully. That includes cmdliner itself:

- [cmdliner](https://github.com/dbuenzli/cmdliner)
- [logs](https://github.com/dbuenzli/logs)
- [fmt](https://github.com/dbuenzli/fmt)
- The "younger" [brr](https://github.com/dbuenzli/brr), which is [planned](https://github.com/aantron/dream/tree/master/example#roadmap) for a full-stack example in Dream.
- [uutf](https://github.com/dbuenzli/uutf)
- etc.

**Comment by aantron** [comment]
And apart from `lib/` being an example that is dangerous to users of Dream with ReScript, if Dream ever does https://github.com/aantron/dream/issues/72 "Port Dream to Node," and wants to also port over some projects from its ecosystem, almost the first thing that will have to be done is that those projects which use `lib/` will have to be adjusted somehow, or hacks will need to be introduced.

**Comment by aantron** [comment]
`src/` also is better for quickly figuring out blame that has to do with the main source (as opposed to tests, etc.) With multiple top-level source directories, you can't immediately do that.

In general, `src/` tends to be tidier, since it groups all of the "meat" of a project apart from all the "other stuff" that a project must also have. Again, you can (and practically all projects do) still separate the libraries and the binaries, etc. `dune` conventions encourage that (and it was always the right thing to do, even back in the day). Project top-level structure with `src/` doesn't grow and mutate over time as a project chooses to drop or support a novel environment (like Lwt, Async, JS, etc.) &mdash; the repo is always tidy, and that growth does not spawn a whole new subdirectory branch of blame to suddenly look at. You can look at hte git history of `src/` and see all of it. Conversely, you can easily miss that there was a `lib_XYZ` directory for a while, if you just didn't know to expect that it was there in 2017-2019.

**Comment by tmattio** [comment]
That's a lot of rather convincing arguments, thank you for taking your time to share 🙂 

There are two issues at hand here I think:

- The use of `lib/` to name a top-level directory
- The use of multiple top-level directories

I'm actually already using subdirectories of `{bin,lib}/`. Typically, I would do:

```
bin/
  - binary_1/
  - binary_2/
  - ...
lib/
  - library_1/
  - library_2/
  - ...
```

So the top-level does not grow. So I think the questions are:

- Should `lib/` be named `src/` instead?
- Should `bin/` be a subdirectory of `src/`, or otherwise named differently.

FWIW Daniel Buenzli's convention seems to be to split the binary code to another top-level directory as well (e.g. [Carcass](https://github.com/dbuenzli/carcass)).

I'm not sure if you're arguing for a single top-level directory containing all of the sources, or if only the prevention of an ever-growing number of top-level directories? Or in other words, does the `src/` and `src-bin/` convention in the Carcass repository seem like a good one to you?

I could get convinced that `src/` is a better convention, but again the most important factor and objective to me is establishing a consistent convention in the ecosystem (I'm not sure if ReScript counts, they seem to inherit conventions more from the JS ecosystem than the OCaml ecosystem).
You've evidently put a lot of thought into this, so maybe there is enough material to open this conversation with the rest of the community?
Would you be willing to ask the question as a Discourse post?

I realized this is a big-time investment for such a question, but the conclusion drawn would potentially make me revisit all of the Spin templates I have created (in addition to the Dream libraries obviously), which will impact a potentially large number of beginners, especially now that Spin is getting more and more recommended.
So I think taking the time to try and gather the community around a convention is worth it.

**Comment by aantron** [comment]
The most important thing is not to use `lib/`. `src-bin/` is strange (again, it's pointless IMO, could use `src/bin`, and if you have two binaries, `src/foo` and `src/bar` is just as flat). But it's not something I would even raise an issue about.

I don't have a goal of getting the whole ecosystem behind this, though I would separately be very happy to see the whole ecosystem move away from `lib/` for new projects. The minimal and justifiable goal is that it is not used with Dream. Dream also doesn't have a goal of following "the community" or waiting for "community" decisions (who is that? how would that thread look? when has "the community" "decided" something?). Which isn't to say that Dream is in any way against the community &mdash; it's just not influenced by considerations of relatively distant projects, nor should they be influenced so directly by Dream.

I would like templaters *used with Dream* to adopt conventions sensible for the users of Dream, however. Such projects have more influence on "the community" at large than one-time discussions anyway, so if you'd like an effect of making the ecosystem more consistent in a global way, it could be indeed advanced through these projects.

If you'd like to poll the community before proceeding with all of Spin (I'm only asking about spin-dream, dream-cli, dream-livereload, etc), you're welcome to open a topic, send me a link, and I will comment with my point of view.

Dream can't ignore ReScript :)

**Comment by aantron** [comment]
Carcass is a good example, though. If you believe it is best to keep `lib` and `bin` at the top level, if you are willing to adopt Carcass convention and call them `src`/`src-bin`, then that will, of course, no longer conflict with ReScript.

**Comment by aantron** [comment]
Another example, in this case where top-level `bin/` for sources is problematic.

As the traditional (and reasonable) name for the *output* directory for binaries (and as used in system directories), I use [`bin/`](https://github.com/aantron/bisect_ppx/tree/binaries) in the root of the Bisect_ppx npm package to store pre-built binaries for Linux and Mac. It would be unfortunate to have to come up with some odd name for this, if I had started with a template that used `bin/`.

There are several other projects in ReScript/Melange that distribute pre-built binaries.

In summary, `bin/` is for staging binaries you are about to install, not for sources of binaries.

Likewise, `lib/` is for staging libraries you are about to install, not for sources of libraries.

One source for the confusion that developed in some corners of OCaml might be that JavaScript and other *non-compiled* languages install *sources* as libraries, hence they have a tendency to put sources in `lib/`. This is inappropriate for OCaml.

**Comment by tmattio** [comment]
Thanks a lot for the additional arguments @aantron!

I haven't personally been impacted by the problems you mentioned, but as I said, I don't have a strong opinion on one or the other. You seem to have quite a good case for not using `lib/`, so let's change the convention. I'll take some time to think of a new one and will update my different repositories (`spin-dream` included) once I've found one I'm happy with. I'm open to your directory structure suggestions in the meantime 🙂

**Comment by aantron** [comment]
BTW I guess it would make sense to move this issue to spin-dream.

**Comment by tmattio** [comment]
I've recently updated the Spin templates to use `src` instead of `lib`. You can find the template here: https://github.com/ocaml-templates/spin-dream

I think we can close this issue then, but don't hesitate to open a new one on the Spin template repo if you have other suggestions/feedback!
